### PR TITLE
Deferred codegen

### PR DESCRIPTION
**Pull request by vchuravy**
```
using Enzyme

import Enzyme: Compiler
import .Compiler.GPUCompiler: deferred_codegen_jobs
@generated function deferred_codegen(::Val{f}, ::Val{tt}) where {f,tt}
    primal, adjoint = Compiler.fspec(f, tt)
    target = Compiler.EnzymeTarget()
    params = Compiler.EnzymeCompilerParams(adjoint, false)
    job    = Compiler.CompilerJob(target, primal, params)

    push!(deferred_codegen_jobs, job)
    id = length(deferred_codegen_jobs)

    quote
        # TODO: add an edge to this method instance to support method redefinitions
        ccall("extern deferred_codegen", llvmcall, Ptr{Cvoid}, (Int,), $id)
    end
end

f(x) = x^3

function test(x)
    ptr = deferred_codegen(Val(f), Val(Tuple{Active{Float64}}))
    ccall(ptr, Float64, (Float64, Float64), x, 1.0)
end
```

### CODE REVIEW

**Comment by vchuravy** [comment]
```julia
using Enzyme

f(x) = x*x^3

function first_order(x)
    x′, = autodiff(f, Active(x))
    x′
end

function second_order(x)
    x′, = autodiff(first_order, Active(x))
    x′
end

@show first_order(1.0)
@show autodiff(first_order, Active(1.0))
@show second_order(1.0)
@show autodiff(second_order, Active(1.0))

@info "Testing GPUs"
using CUDA
A = CUDA.zeros(2)

function kernel(A,x)
    A[1] = first_order(x)
    A[2] = second_order(x)
    nothing
end

@show A
```

**Comment by vchuravy** [comment]
Should use https://github.com/JuliaGPU/GPUCompiler.jl/blob/3aa87d65852aea91cb0d942b1110008961966ca7/src/irgen.jl#L296-L327 to implement the ABI

File: src/compiler/validation.jl [hunk 1]
```diff
@@ -95,52 +95,75 @@ function check_ir!(job, errors, inst::LLVM.CallInst)
...
-    #                 fn, file, line, linfo, fromC, inlined = last(frames)
-    #             else
-    #                 fn, file, line, linfo, fromC, inlined, ip = last(frames)
-    #             end
-    #             push!(errors, (POINTER_FUNCTION, bt, fn))
-    #         else
-    #             push!(errors, (POINTER_FUNCTION, bt, nothing))
-    #         end
-    #     end
+        # detect calls to literal pointers
```
Comment (vchuravy) [inline src/compiler/validation.jl:105]:
```suggestion
        # detect calls to literal pointers and de-optimize them
```

File: test/runtests.jl [hunk 2]
```diff
@@ -148,46 +150,45 @@ end
...
     @test fd ≈ ForwardDiff.derivative(foo, x)
     @test fd ≈ Zygote.gradient(foo, x)[1]
-    @test fd ≈ autodiff(foo, Active(x))
+    @test fd ≈ first(autodiff(foo, Active(x)))
 end
 
+"""
+    J(ν, z) := ∑ (−1)^k / Γ(k+1) / Γ(k+ν+1) * (z/2)^(ν+2k)
+"""
+function besselj(ν, z, atol=1e-8)
```
Comment (vchuravy) [inline test/runtests.jl:166]: Rename so that we don't have a name conflict with SpecialFunctions

File: Project.toml [hunk 3]
```diff
@@ -4,17 +4,25 @@ authors = ["William Moses <wmoses@mit.edu>", "Valentin Churavy <vchuravy@mit.edu
 version = "0.4.1"
 
 [deps]
+Adapt = "79e6a3ab-5dfb-504d-930d-738a2a938a0e"
```
Comment (vchuravy) [inline Project.toml:7]: Can we move the adapt rules from the test case into this Enzyme.jl

File: Project.toml [hunk 4]
```diff
@@ -4,17 +4,25 @@ authors = ["William Moses <wmoses@mit.edu>", "Valentin Churavy <vchuravy@mit.edu
...
 CEnum = "fa961155-64e5-5f13-b03f-caf6b980ea82"
+CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
+CUDAKernels = "72cfdca4-0801-4ab0-bf6a-d52aa10adc57"
 Cassette = "7057c7e9-c182-5462-911a-8362d720325c"
+Cthulhu = "f68482b8-f384-11e8-15f7-abe071a5a75f"
```
Comment (vchuravy) [inline Project.toml:12]:
```suggestion

```

File: Project.toml [hunk 5]
```diff
@@ -4,17 +4,25 @@ authors = ["William Moses <wmoses@mit.edu>", "Valentin Churavy <vchuravy@mit.edu
...
 Enzyme_jll = "7cc45869-7501-5eee-bdea-0790c847d4ef"
 GPUCompiler = "61eb1bfa-7361-4325-ad38-22787b887f55"
+KernelAbstractions = "63c18a36-062a-441e-b654-da1e3ab1ce7c"
```
Comment (vchuravy) [inline Project.toml:15]:
```suggestion

```

File: Project.toml [hunk 6]
```diff
@@ -4,17 +4,25 @@ authors = ["William Moses <wmoses@mit.edu>", "Valentin Churavy <vchuravy@mit.edu
...
 LLVM = "929cbde3-209d-540e-8aea-75f648917ca0"
 Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
+ObjectFile = "d8793406-e978-5875-9003-1fc021f44a92"
+OpenSpecFun_jll = "efe28fd5-8261-553b-a9e1-b2916fc3738e"
```
Comment (vchuravy) [inline Project.toml:19]:
```suggestion

```

File: Project.toml [hunk 7]
```diff
@@ -4,17 +4,25 @@ authors = ["William Moses <wmoses@mit.edu>", "Valentin Churavy <vchuravy@mit.edu
...
+Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
```
Comment (vchuravy) [inline Project.toml:20]: What is `Statistics` used for?

File: Project.toml [hunk 8]
```diff
@@ -4,17 +4,25 @@ authors = ["William Moses <wmoses@mit.edu>", "Valentin Churavy <vchuravy@mit.edu
 version = "0.4.1"
 
 [deps]
+Adapt = "79e6a3ab-5dfb-504d-930d-738a2a938a0e"
 CEnum = "fa961155-64e5-5f13-b03f-caf6b980ea82"
+CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
```
Comment (vchuravy) [inline Project.toml:9]:
```suggestion

```

File: Project.toml [hunk 9]
```diff
@@ -4,17 +4,25 @@ authors = ["William Moses <wmoses@mit.edu>", "Valentin Churavy <vchuravy@mit.edu
...
+CUDAKernels = "72cfdca4-0801-4ab0-bf6a-d52aa10adc57"
```
Comment (vchuravy) [inline Project.toml:10]:
```suggestion

```

File: Project.toml [hunk 10]
```diff
@@ -4,17 +4,25 @@ authors = ["William Moses <wmoses@mit.edu>", "Valentin Churavy <vchuravy@mit.edu
...
 Enzyme_jll = "7cc45869-7501-5eee-bdea-0790c847d4ef"
 GPUCompiler = "61eb1bfa-7361-4325-ad38-22787b887f55"
+KernelAbstractions = "63c18a36-062a-441e-b654-da1e3ab1ce7c"
 LLVM = "929cbde3-209d-540e-8aea-75f648917ca0"
 Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
+ObjectFile = "d8793406-e978-5875-9003-1fc021f44a92"
+OpenSpecFun_jll = "efe28fd5-8261-553b-a9e1-b2916fc3738e"
+Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
 
 [compat]
```
Comment (vchuravy) [inline Project.toml:22]: We will need compat bounds for the new dependencies

File: src/compiler.jl [hunk 11]
```diff
@@ -79,58 +83,127 @@ include("compiler/cassette.jl")
...
+    # end
+
     args_activity     = API.CDIFFE_TYPE[]
     uncacheable_args  = Bool[]
     args_typeInfo     = TypeTree[]
     args_known_values = API.IntList[]
 
+    ctx = LLVM.context(mod)
+
+    # @show mod, tt
```
Comment (vchuravy) [inline src/compiler.jl:169]: rm

File: src/compiler.jl [hunk 12]
```diff
@@ -79,58 +83,127 @@ include("compiler/cassette.jl")
...
     for T in tt
+        source_typ = eltype(T)
+        if GPUCompiler.isghosttype(source_typ) || Core.Compiler.isconstType(source_typ)
+            @assert T <: Const
+            continue
+        end
+        isboxed = GPUCompiler.deserves_argbox(source_typ)
+        # @show source_typ, isboxed
+
+        # if !isboxed && nameof(source_typ.name.module) != :CUDA
```
Comment (vchuravy) [inline src/compiler.jl:183]: rm

File: src/compiler.jl [hunk 13]
```diff
@@ -189,21 +268,641 @@ function enzyme!(mod, primalf, adjoint, rt, split)
...
+            # uncacheable_args, #=augmented=#C_NULL, #=atomicAdd=# parallel, #=postOpt=#false))
+            uncacheable_args, #=augmented=#C_NULL, #=atomicAdd=# parallel, #=postOpt=#false))
         augmented_primalf = nothing
     end
-    
     return adjointf, augmented_primalf
 end
 
-include("compiler/thunk.jl")
+function Base.in(attr::LLVM.EnumAttribute, iter::LLVM.FunctionAttrSet)
```
Comment (vchuravy) [inline src/compiler.jl:284]: upstream to LLVM.jl?

File: src/compiler.jl [hunk 14]
```diff
@@ -189,21 +268,641 @@ function enzyme!(mod, primalf, adjoint, rt, split)
...
+        end
+    end
+    return false
+end
+
+# Modified from GPUCompiler/src/irgen.jl:365 lower_byval
+function lower_convention(@nospecialize(job::CompilerJob), mod::LLVM.Module, entry_f::LLVM.Function)
+    ctx = context(mod)
+    entry_ft = eltype(llvmtype(entry_f)::LLVM.PointerType)::LLVM.FunctionType
+    # @compiler_assert return_type(entry_ft) == LLVM.VoidType(ctx) job
```
Comment (vchuravy) [inline src/compiler.jl:305]: rb

File: src/compiler.jl [hunk 15]
```diff
@@ -189,21 +268,641 @@ function enzyme!(mod, primalf, adjoint, rt, split)
...
+                continue
+            end
+            expr = argexprs[i]
+
+            isboxed = GPUCompiler.deserves_argbox(source_typ)
+            llvmT = isboxed ? T_prjlvalue : convert(LLVMType, source_typ, ctx)
+            argexpr = Expr(:., expr, QuoteNode(:val))
+            if isboxed
+                push!(types, Any)
+            # elseif isa(llvmT, LLVM.SequentialType) && nameof(source_typ.name.module) != :CUDA # et->isAggregateType
```
Comment (vchuravy) [inline src/compiler.jl:544]: rb

File: src/compiler.jl [hunk 16]
```diff
@@ -189,21 +268,641 @@ function enzyme!(mod, primalf, adjoint, rt, split)
...
+                push!(sret_types, source_typ)
+                push!(T_JuliaSRet, llvmT)
+                if !isboxed # XXX: Not consistent
+                    push!(T_EnzymeSRet, llvmT)
+                end
+            elseif T <: Duplicated || T <: DuplicatedNoNeed
+                argexpr =  Expr(:., expr, QuoteNode(:dval))
+                if isboxed
+                    push!(types, Any)
+#                elseif isa(llvmT, LLVM.SequentialType) && nameof(source_typ.name.module) != :CUDA# et->isAggregateType
```
Comment (vchuravy) [inline src/compiler.jl:572]: rm

File: src/compiler/validation.jl [hunk 17]
```diff
@@ -13,28 +14,35 @@ function check_ir(job, args...)
 end
 
 function check_ir!(job, errors, mod::LLVM.Module)
-    for f in functions(mod)
-        check_ir!(job, errors, f)
+    # @show "premod", mod
```
Comment (vchuravy) [inline src/compiler/validation.jl:17]: rm

File: src/compiler/validation.jl [hunk 18]
```diff
@@ -95,22 +103,189 @@ function check_ir!(job, errors, inst::LLVM.CallInst)
...
+                        end
+                    end
+                    for g in globals(inmod)
+                        linkage!(g, LLVM.API.LLVMExternalLinkage)
+                    end
+                    # override libdevice's triple and datalayout to avoid warnings
+                    triple!(inmod, triple(mod))
+                    datalayout!(inmod, datalayout(mod))
+                    # @show mod
+                    # @show inmod
```
Comment (vchuravy) [inline src/compiler/validation.jl:232]: rm

File: test/abi.jl [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    res0, = autodiff(unused, Const(nothing), Active(2.0))
+    @test res0 ≈ 1.0
+
+    # Multi arg => sret
+    mul(x, y) = x * y
+    pair = autodiff(mul, Active(2.0), Active(3.0))
+    @test pair[1] ≈ 3.0
+    @test pair[2] ≈ 2.0
+
+    # SeqeuntialType
```
Comment (vchuravy) [inline test/abi.jl:30]:
```suggestion
    # SequentialType
```

File: src/compiler/validation.jl [hunk 20]
```diff
@@ -95,22 +100,182 @@ function check_ir!(job, errors, inst::LLVM.CallInst)
...
+            end
+            if isa(fname, LLVM.ConstantArray) && eltype(llvmtype(fname)) == LLVM.IntType(8, ctx)
+                fname = String(map((x)->convert(UInt8, x), collect(fname)[1:(end-1)]))
+            end
+            hnd = LLVM.Value(LLVM.LLVM.API.LLVMGetOperand(inst, 2))
+            if isa(hnd, LLVM.GlobalVariable)
+                hnd = LLVM.name(hnd)
+            end
+
+            @show flib, fname, hnd
```
Comment (giordano) [inline src/compiler/validation.jl:151]: Leftover?

File: src/compiler/validation.jl [hunk 21]
```diff
@@ -1,4 +1,5 @@
 using LLVM
+using ObjectFile
```
Comment (giordano) [inline src/compiler/validation.jl:2]: What is `ObjectFile` used for?

**Pull request merged**